rust/template/d3log/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+
+            let (_p, _init_batch, ep, _dispatch, forwarder) = async_error!(
+                self.eval,
+                start_instance(self.rt.clone(), self.new_evaluator.clone(), uuid)
+            );
+
+            // there is a race (missing events, duplicated events) because we dont have
+            // a sequence number here
+            let b = { Batch::Value(self.accumulator.lock().expect("lock").clone()) };
+            ep.send(b);
+
+            self.forwarder.register(uuid, ep.clone());
+            forwarder.register(self.eval.clone().myself(), self.evalport.clone());
+            let threads: u64 = 1;
+            let bytes: u64 = 1;
+
+            self.broadcast.send(fact!(d3_application::InstanceStatus,
+                                      time => self.eval.clone().now().into_record(),
+                                      id => uuid.into_record(),
+                                      memory_bytes => bytes.into_record(),
+                                      threads => threads.into_record()));
+        }
+    }
+}
+
+struct DebugPort {
+    eval: Evaluator,
+}
+
+impl Transport for DebugPort {
+    fn send(&self, b: Batch) {
+        for (_r, f, w) in &RecordBatch::from(self.eval.clone(), b) {
+            println!("{} {}", f, w);
+        }
+    }
+}
+
+pub fn start_instance(
+    rt: Arc<Runtime>,
+    new_evaluator: Arc<dyn Fn(Node, Port) -> Result<(Evaluator, Batch), Error> + Send + Sync>,
+    uuid: u128,
+) -> Result<(Port, Batch, Port, Port, Arc<Forwarder>), Error> {
+    let broadcast = Broadcast::new();
+    let (eval, init_batch) = new_evaluator(uuid, broadcast.clone())?;
+    let dispatch = Arc::new(Dispatch::new(eval.clone()));
+
+    broadcast
+        .clone()
+        .subscribe(Arc::new(DebugPort { eval: eval.clone() }));
+    let forwarder = Forwarder::new(eval.clone(), dispatch.clone(), broadcast.clone());
+    let accu_batch = Arc::new(Mutex::new(DDValueBatch::new()));
+
+    // shouldn't evaluator just implement Transport?
+    let eval_port = Arc::new(EvalPort {
+        forwarder: forwarder.clone(),
+        dispatch: dispatch.clone(),
+        eval: eval.clone(),
+        queue: Arc::new(Mutex::new(VecDeque::new())),
+    });
+
+    dispatch.clone().register(
+        "d3_application::ThreadInstance",
+        Arc::new(ThreadInstance {
+            rt: rt.clone(),
+            accumulator: accu_batch.clone(),
+            eval: eval.clone(),
+            evalport: eval_port.clone(),
+            forwarder: forwarder.clone(),
+            new_evaluator: new_evaluator.clone(),
+            broadcast: broadcast.clone(),
+        }),
+    )?;
+
+    broadcast.clone().subscribe(eval_port.clone());
+    broadcast.clone().subscribe(Arc::new(AccumulatePort {
+        eval: eval.clone(),
+        b: accu_batch.clone(),
+    }));
+    eval_port.send(fact!(d3_application::Myself, me => uuid.into_record()));
+    Ok((broadcast, init_batch, eval_port, dispatch, forwarder))
```

Comment:
I'm going to go ahead and make an Instance structure so that we don't have to carry around the 8 or 9 objects explicitly. these structures tend to collect junk, but its just a lot of noise right now. plus some of the giant blocks of clones can just become Instance.clone()